################################################################################
SHORT DESCRIPTION: 
################################################################################
Add vhd icbinn support.

################################################################################
LONG DESCRIPTION: 
################################################################################
Add vhd icbinn support.

################################################################################
CHANGELOG 
################################################################################

################################################################################
REMOVAL 
################################################################################

################################################################################
UPSTREAM PLAN
################################################################################
Not sure.

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES 
################################################################################
--- a/drivers/Makefile.am
+++ b/drivers/Makefile.am
@@ -7,21 +7,22 @@ AM_CFLAGS += $(if $(GCOV),-fprofile-dir=
 
 AM_CPPFLAGS  = -D_GNU_SOURCE
 AM_CPPFLAGS += -I$(top_srcdir)/include
+ICBLIBS = -licbinn -lpthread -ldl
 
 libexec_PROGRAMS = tapdisk
 
 tapdisk_SOURCES = tapdisk2.c
-tapdisk_LDADD = libtapdisk.la
+tapdisk_LDADD = libtapdisk.la $(ICBLIBS)
 
 noinst_PROGRAMS = tapdisk-stream
 
-tapdisk_stream_LDADD = libtapdisk.la
+tapdisk_stream_LDADD = libtapdisk.la $(ICBLIBS)
 
 sbin_PROGRAMS  = td-util
 sbin_PROGRAMS += td-rated
 
 td_util_SOURCES = td.c
-td_util_LDADD = libtapdisk.la
+td_util_LDADD = libtapdisk.la $(ICBLIBS)
 
 noinst_LTLIBRARIES = libtapdisk.la
 
--- a/include/libvhd.h
+++ b/include/libvhd.h
@@ -127,9 +127,20 @@ typedef struct vhd_bat             vhd_b
 typedef struct vhd_batmap          vhd_batmap_t;
 typedef struct dd_batmap_hdr       vhd_batmap_header_t;
 typedef struct prt_loc             vhd_parent_locator_t;
+typedef struct vhd_devops          vhd_devops_t;
 typedef struct vhd_context         vhd_context_t;
 typedef uint32_t                   vhd_flag_creat_t;
 
+struct vhd_devops {
+	off_t (*position)          (vhd_context_t *);
+	int (*seek)                (vhd_context_t *, off64_t, int);
+	int (*read)                (vhd_context_t *, void *, size_t);
+	int (*write)               (vhd_context_t *, void *, size_t);
+	int (*pread)               (vhd_context_t *, void *, size_t, off64_t);
+	int (*pwrite)              (vhd_context_t *, void *, size_t, off64_t);
+	void (*close)              (vhd_context_t *);
+};
+
 struct vhd_bat {
 	uint32_t                   spb;
 	uint32_t                   entries;
@@ -157,6 +168,7 @@ struct vhd_context {
 	vhd_footer_t               footer;
 	vhd_bat_t                  bat;
 	vhd_batmap_t               batmap;
+	vhd_devops_t		  *devops;
 
 	struct crypto_blkcipher   *xts_tfm;
 	struct list_head           next;
@@ -375,4 +387,8 @@ int vhd_io_write_bytes(vhd_context_t *,
 
 /** @} */ // End of LibVHD Doxygen group
 
+char *vhd_canonpath(const char *path, char *resolved_path, size_t size);
+struct ICBINN_struct * vhd_icbinn_key(void);
+struct ICBINN_struct * vhd_icbinn_vhd(void);
+
 #endif
--- a/vhd/Makefile.am
+++ b/vhd/Makefile.am
@@ -12,7 +12,7 @@ bin_PROGRAMS  = vhd-util
 bin_PROGRAMS += vhd-index
 bin_PROGRAMS += vhd-update
 
-LDADD = lib/libvhd.la -luuid
+LDADD = lib/libvhd.la -luuid -Llib -licbinn -ldl -lpthread
 
 vhd_index_LDADD = lib/libvhd.la -luuid
 
--- a/vhd/lib/Makefile.am
+++ b/vhd/lib/Makefile.am
@@ -42,7 +42,7 @@ libvhd_la_SOURCES += xattr.h
 
 libvhd_la_LDFLAGS = -version-info 1:1:1
 
-libvhd_la_LIBADD = -luuid -ldl $(LIBICONV)  $(top_srcdir)/lvm/liblvmutil.la
+libvhd_la_LIBADD = -luuid -lpthread -licbinn -ldl $(LIBICONV) $(top_srcdir)/lvm/liblvmutil.la
 
 if ENABLE_TESTS
 MAYBE_test = test
--- /dev/null
+++ b/vhd/lib/icbinn.c
@@ -0,0 +1,334 @@
+/*
+ * Copyright (c) 2012 Citrix Systems, Inc.
+ */
+
+#include <icbinn.h>
+
+static ICBINN *icb_vhd;
+
+static ICBINN *icb_key;
+
+ICBINN *
+vhd_icbinn_key (void)
+{
+  char proto[1024], *ptr, *host;
+  int port;
+
+  if (icb_key)
+    return icb_key;
+
+  ptr = getenv ("LIBVHD_ICBINN_KEY_SERVER");
+  if (!ptr)
+    return NULL;
+
+  strncpy (proto, ptr, sizeof (proto));
+  proto[sizeof (proto) - 1] = 0;
+
+  host = index (proto, ':');
+  if (!host)
+    return NULL;
+  *(host++) = 0;
+
+  ptr = index (host, ':');
+  if (!ptr)
+    {
+      port = ICBINN_PORT;
+    }
+  else
+    {
+      *ptr = 0;
+      port = atoi (ptr + 1);
+    }
+
+  if (!strcmp (proto, "argo"))
+    {
+      icb_key = icbinn_clnt_create_argo (atoi (host), port);
+    }
+  else if (!strcmp (proto, "tcp"))
+    {
+      icb_key = icbinn_clnt_create_tcp (host, port);
+    }
+
+  return icb_key;
+}
+
+ICBINN *
+vhd_icbinn_vhd (void)
+{
+  char proto[1024], *ptr, *host;
+  int port;
+
+  if (icb_vhd)
+    return icb_vhd;
+
+  ptr = getenv ("LIBVHD_ICBINN_VHD_SERVER");
+  if (!ptr)
+    return NULL;
+
+  strncpy (proto, ptr, sizeof (proto));
+  proto[sizeof (proto) - 1] = 0;
+
+  host = index (proto, ':');
+  if (!host)
+    return NULL;
+  *(host++) = 0;
+
+  ptr = index (host, ':');
+  if (!ptr)
+    {
+      port = ICBINN_PORT;
+    }
+  else
+    {
+      *ptr = 0;
+      port = atoi (ptr + 1);
+    }
+
+  if (!strcmp (proto, "argo"))
+    {
+      icb_vhd = icbinn_clnt_create_argo (atoi (host), port);
+    }
+  else if (!strcmp (proto, "tcp"))
+    {
+      icb_vhd = icbinn_clnt_create_tcp (host, port);
+    }
+
+  return icb_vhd;
+}
+
+static off_t
+vhd_icbinn_devops_position (vhd_context_t * vhd)
+{
+  return (off_t) vhd->offset;
+}
+
+static int
+vhd_icbinn_devops_seek (vhd_context_t * vhd, off64_t off, int whence)
+{
+  struct icbinn_stat buf;
+
+  switch (whence)
+    {
+      //XXX: both of these ought to check for seeks beyond EOF, but they don't
+    case SEEK_SET:
+      vhd->offset = off;
+      break;
+    case SEEK_CUR:
+      vhd->offset += off;
+      break;
+    case SEEK_END:
+
+      if (icbinn_stat (icb_vhd, vhd->file, &buf))
+        return -1;
+
+      if (buf.type != ICBINN_TYPE_FILE)
+        return -1;
+
+      vhd->offset = buf.size + off;
+      break;
+    default:
+      return -1;
+    }
+
+  return 0;
+}
+
+static int
+vhd_icbinn_devops_read (vhd_context_t * vhd, void *buf, size_t count)
+{
+  size_t ret;
+
+  ret = icbinn_pread (icb_vhd, vhd->fd, buf, count, vhd->offset);
+
+  if (ret > 0)
+    vhd->offset += ret;
+
+  return (ret == count) ? 0 : -EIO;
+}
+
+static int
+vhd_icbinn_devops_pread (vhd_context_t * vhd,
+                         void *buf, size_t size, off64_t off)
+{
+  size_t ret = icbinn_pread (icb_vhd, vhd->fd, buf, size, off);
+
+  return (ret == size) ? 0 : -EIO;
+}
+
+static int
+vhd_icbinn_devops_write (vhd_context_t * vhd, void *buf, size_t count)
+{
+  size_t ret;
+
+  ret = icbinn_pwrite (icb_vhd, vhd->fd, buf, count, vhd->offset);
+
+  if (ret > 0)
+    vhd->offset += ret;
+
+  return (ret == count) ? 0 : -EIO;
+}
+
+static int
+vhd_icbinn_devops_pwrite (vhd_context_t * vhd,
+                          void *buf, size_t size, off64_t off)
+{
+  size_t ret = icbinn_pwrite (icb_vhd, vhd->fd, buf, size, off);
+
+  return (ret == size) ? 0 : -EIO;
+}
+
+static void
+vhd_icbinn_devops_close (vhd_context_t * vhd)
+{
+  char *file = vhd->file;
+  int fd = vhd->fd;
+
+  vhd->fd = -1;
+  __vhd_close (vhd);
+
+  if (file)
+    icbinn_close (icb_vhd, fd);
+}
+
+static vhd_devops_t vhd_icbinn_devops = {
+  .position = vhd_icbinn_devops_position,
+  .seek = vhd_icbinn_devops_seek,
+  .read = vhd_icbinn_devops_read,
+  .write = vhd_icbinn_devops_write,
+  .pread = vhd_icbinn_devops_pread,
+  .pwrite = vhd_icbinn_devops_pwrite,
+  .close = vhd_icbinn_devops_close,
+};
+
+
+
+/* Shamelessly stolen of glibc 2.9 with modifications */
+
+/* Return the canonical absolute name of file NAME.  A canonical name
+   does not contain any `.', `..' components nor any repeated path
+   separators ('/') or symlinks.  All path components must exist.  If
+   RESOLVED is null, the result is malloc'd; otherwise, if the
+   canonical name is PATH_MAX chars or more, returns null with `errno'
+   set to ENAMETOOLONG; if the name fits in fewer than PATH_MAX chars,
+   returns the name in RESOLVED.  If the name cannot be resolved and
+   RESOLVED is non-NULL, it contains the path of the first component
+   that cannot be resolved.  If the path can be resolved, RESOLVED
+   holds the same value as the value returned.  */
+
+char *
+vhd_canonpath(const char *name, char *resolved, size_t size)
+{
+  char *rpath, *dest;
+  const char *start, *end, *rpath_limit;
+  long int path_max;
+
+  if (!vhd_icbinn_vhd ())
+    return canonpath (name, resolved, size);
+
+  if ((name == NULL) || (name[0] == '\0'))
+    return canonpath (name, resolved, size);
+
+#ifdef PATH_MAX
+  path_max = PATH_MAX;
+#else
+  path_max = pathconf (name, _PC_PATH_MAX);
+  if (path_max <= 0)
+    path_max = 1024;
+#endif
+
+  if (resolved == NULL)
+    {
+      rpath = malloc (path_max);
+      if (rpath == NULL)
+        return NULL;
+    }
+  else
+    rpath = resolved;
+  rpath_limit = rpath + path_max;
+
+  rpath[0] = '/';
+  dest = rpath + 1;
+
+  for (start = end = name; *start; start = end)
+    {
+      struct icbinn_stat st;
+
+      /* Skip sequence of multiple path-separators.  */
+      while (*start == '/')
+        ++start;
+
+      /* Find end of path component.  */
+      for (end = start; *end && *end != '/'; ++end)
+        /* Nothing.  */ ;
+
+      if (end - start == 0)
+        break;
+      else if (end - start == 1 && start[0] == '.')
+        /* nothing */ ;
+      else if (end - start == 2 && start[0] == '.' && start[1] == '.')
+        {
+          /* Back up to previous component, ignore if at root already.  */
+          if (dest > rpath + 1)
+            while ((--dest)[-1] != '/');
+        }
+      else
+        {
+          size_t new_size;
+
+          if (dest[-1] != '/')
+            *dest++ = '/';
+
+          if (dest + (end - start) >= rpath_limit)
+            {
+              ptrdiff_t dest_offset = dest - rpath;
+              char *new_rpath;
+
+              if (resolved)
+                {
+                  errno = ENAMETOOLONG;
+                  if (dest > rpath + 1)
+                    dest--;
+                  *dest = '\0';
+                  goto error;
+                }
+              new_size = rpath_limit - rpath;
+              if (end - start + 1 > path_max)
+                new_size += end - start + 1;
+              else
+                new_size += path_max;
+              new_rpath = (char *) realloc (rpath, new_size);
+              if (new_rpath == NULL)
+                goto error;
+              rpath = new_rpath;
+              rpath_limit = rpath + new_size;
+
+              dest = rpath + dest_offset;
+            }
+
+          dest = mempcpy (dest, start, end - start);
+          *dest = '\0';
+
+          if (icbinn_stat (icb_vhd, rpath, &st) < 0)
+            {
+              errno = ENOENT;
+              goto error;
+            }
+
+          if ((st.type != ICBINN_TYPE_DIRECTORY) && (*end != '\0'))
+            {
+              errno = ENOTDIR;
+              goto error;
+            }
+        }
+    }
+  if (dest > rpath + 1 && dest[-1] == '/')
+    --dest;
+  *dest = '\0';
+
+  return rpath;
+
+error:
+  if (resolved == NULL)
+    free (rpath);
+  return NULL;
+}
--- a/vhd/lib/libvhd.c
+++ b/vhd/lib/libvhd.c
@@ -50,6 +50,7 @@
 #include <sys/stat.h>
 #include <sys/sysmacros.h>
 #include <sys/types.h>
+#include <stddef.h>
 
 #include "debug.h"
 #include "xattr.h"
@@ -103,6 +104,9 @@ static int vhd_cache_enabled(vhd_context
 static int vhd_cache_load(vhd_context_t *);
 static int vhd_cache_unload(vhd_context_t *);
 static vhd_context_t * vhd_cache_get_parent(vhd_context_t *);
+static void __vhd_close(vhd_context_t *ctx);
+
+#include "icbinn.c"
 
 static inline int
 old_test_bit(volatile char *addr, int nr)
@@ -1484,7 +1488,7 @@ vhd_find_parent(vhd_context_t *ctx, cons
 	}
 
 	/* check parent path relative to child's directory */
-	cpath = canonpath(ctx->file, __cpath, sizeof(__cpath));
+	cpath = vhd_canonpath(ctx->file, __cpath, sizeof(__cpath));
 	if (!cpath) {
 		err = -errno;
 		goto out;
@@ -1498,7 +1502,7 @@ vhd_find_parent(vhd_context_t *ctx, cons
 	}
 
 	if (!access(location, R_OK)) {
-		path = canonpath(location, __location, sizeof(__location));
+		path = vhd_canonpath(location, __location, sizeof(__location));
 		if (path) {
 			*_location = strdup(path);
 			if (*_location)
@@ -1874,12 +1878,12 @@ vhd_parent_locator_write_at(vhd_context_
 			    const char *parent, off64_t off, uint32_t code,
 			    size_t max_bytes, vhd_parent_locator_t *loc)
 {
-	struct stat stats;
 	int err, len, size;
 	char *absolute_path, *relative_path, *encoded;
 	char __parent[PATH_MAX];
 	void *block;
 	int (*encode_location_fn)(char*, char**, int*);
+	ICBINN *icb = vhd_icbinn_vhd();
 
 	memset(loc, 0, sizeof(vhd_parent_locator_t));
 
@@ -1911,6 +1915,7 @@ vhd_parent_locator_write_at(vhd_context_
 		goto out;
 	}
 
+	struct stat stats;
 	err = stat(absolute_path, &stats);
 	if (err) {
 		err = -errno;
--- a/vhd/lib/relative-path.c
+++ b/vhd/lib/relative-path.c
@@ -38,6 +38,8 @@
 #include <string.h>
 #include <limits.h>
 
+char *vhd_canonpath(const char *path, char *resolved_path, size_t size);
+
 #include "relative-path.h"
 #include "canonpath.h"
 
@@ -240,14 +242,14 @@ relative_path_to(char *from, char *to, i
 		return NULL;
 	}
 
-	to_absolute = canonpath(to, __to_absolute, sizeof(__to_absolute));
+	to_absolute = vhd_canonpath(to, __to_absolute, sizeof(__to_absolute));
 	if (!to_absolute) {
 		EPRINTF("failed to get absolute path of %s\n", to);
 		*err = -errno;
 		goto out;
 	}
 
-	from_absolute = canonpath(from, __from_absolute, sizeof(__from_absolute));
+	from_absolute = vhd_canonpath(from, __from_absolute, sizeof(__from_absolute));
 	if (!from_absolute) {
 		EPRINTF("failed to get absolute path of %s\n", from);
 		*err = -errno;
--- a/vhd/lib/vhd-util-key.c
+++ b/vhd/lib/vhd-util-key.c
@@ -37,6 +37,7 @@
 #include <sys/stat.h>
 #include <dlfcn.h>
 
+#include "icbinn.h"
 #include "libvhd.h"
 
 #define MAX_KEY_SIZE 512
@@ -178,18 +179,76 @@ out:
 }
 
 static int
+vhd_util_read_key_icbinn(ICBINN *icb, const char *keypath, uint8_t *key,
+		  size_t max_size, size_t *out_size)
+{
+	int fd = -1, err;
+	ssize_t size;
+	struct icbinn_stat sb;
+
+
+
+	err = icbinn_stat(icb, keypath, &sb);
+	if (err) {
+		ERR("failed to stat %s\n", keypath);
+		err = -ENOENT;
+		goto out;
+	}
+
+	if (sb.type != ICBINN_TYPE_FILE) {
+		ERR("%s isn't a file\n", keypath);
+		err = -EISDIR;
+		goto out;
+	}
+
+	fd = icbinn_open(icb, keypath, O_RDONLY);
+	if (fd == -1) {
+		ERR("failed to open %s\n", keypath );
+		err = -EINVAL;
+		goto out;
+	}
+
+	size = icbinn_pread(icb, fd, key, max_size, 0);
+	if (size == -1) {
+		ERR("failed to read key\n");
+		err = -EIO;
+		goto out;
+	}
+
+	if (size != sb.size) {
+		ERR("short read of key\n");
+		err = -EIO;
+		goto out;
+	}
+
+	if (out_size)
+		*out_size = size;
+
+out:
+	if (fd != -1)
+		icbinn_close(icb,fd);
+	return err;
+}
+
+static int
 vhd_util_calculate_keyhash(struct vhd_keyhash *keyhash, const char *keypath, size_t key_bytes)
 {
 	int err;
 	size_t read_bytes;
 	uint8_t key[MAX_KEY_SIZE/8];
+	ICBINN *icb = vhd_icbinn_key();
 
 	if (key_bytes == 0)
 		read_bytes = MAX_KEY_SIZE/8;
 	else
 		read_bytes = key_bytes;
 
-	err = vhd_util_read_key(keypath, key, &read_bytes);
+	if (icb) {
+		err = vhd_util_read_key_icbinn(icb, keypath, key, read_bytes, &read_bytes);
+	}
+	else {
+		err = vhd_util_read_key(keypath, key, &read_bytes);
+	}
 	if (err) {
 		ERR("failed to read key: %d\n", err);
 		goto out;
--- a/vhd/lib/vhd-util-scan.c
+++ b/vhd/lib/vhd-util-scan.c
@@ -787,7 +787,7 @@ vhd_util_scan_open(vhd_context_t *vhd, s
 	else {
 		char __image_name[PATH_MAX];
 
-		image->name = canonpath(target->name, __image_name, sizeof(__image_name));
+		image->name = vhd_canonpath(target->name, __image_name, sizeof(__image_name));
 		if (image->name)
 			image->name = strdup(__image_name);
 		if (!image->name) {
--- a/vhd/lib/vhd-util-snapshot.c
+++ b/vhd/lib/vhd-util-snapshot.c
@@ -174,7 +174,7 @@ vhd_util_snapshot(int argc, char **argv)
 		goto usage;
 	}
 
-	ppath = canonpath(pname, __ppath, sizeof(__ppath));
+	ppath = vhd_canonpath(pname, __ppath, sizeof(__ppath));
 	if (!ppath)
 		return -errno;
 
